keps/sig-cli/20190920-future-of-kubectl-cp.md

```diff
@@ -0,0 +1,141 @@
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+This document summarizes and originates from this email thread, 
+[Proposal to drop kubectl cp](https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/kubernetes-sig-cli/_zUy67lK49k/aE6vncYiAgAJ).   
+
+This document aims to solidify the future of `kubectl cp` as a tool that provides basic function of copying files between local environments and pods.  Any advanced use cases
+such as those involving symlinks or modifying file permissions should be performed outside of `kubectl cp` through `kubectl exec`, addons, or shell commands.    
+
+Over the past few releases, there have been numerous security issues with `kubectl cp` that have resulted in release updates in all supported versions of kubectl.
+At the same time,any new PR that extends `kubectl cp` must undergo extra reviews to evaluate security threats that may arise [1][2].  Over the past few months,
+security fixes have required dropping edge cases and function of the command.  It is increasingly difficult to maintain a cp command that is both
+useful and secure.  There are alternative approaches that provide the same function as `kubectl cp` that are proven more secure [3].  Using `standalone pipe + tar`
+provides transparency when copying files as well as more proven mitigations for path traversals, symlink directory escapes, tar bombs, and other exploits.
+Use of tar is also more more featureful, in that it can preserve file permissions and copy pod-to-pod.   
+
+With all of this in mind the cost and risk of maintaining the cp command should be weighed against what is considered crucial functionality in kubectl. 
+It's better to address 80% of use cases with a simple tool than trying to address the remaining 20% at the cost of risking those 80%.     
+
+[1] https://github.com/kubernetes/kubernetes/pull/78622   
+[2] https://github.com/kubernetes/kubernetes/pull/73053   
+[3] https://gist.github.com/tallclair/9217e2694b5fdf27b55d6bd1fda01b53   
+
+## Motivation
+
+- The `kubectl cp` command has been the subject of multiple security vulnerability reports.
+- Even with basic use cases of `kubectl cp`, there are long-standing issues that have not been addressed.   
+     1) container images are required to have the tar binary in order to use `kubectl cp`.    
+     2) cp command is not well maintained for Windows, if it works at all.    
```

Comment:
What is the evidence for this statement?

Reply:
good point, b4 i make a statement like that i should do my research (i took that point from the email thread w/out verifying) will do..

Reply:
hm, this is still open, although there is an open PR: https://github.com/kubernetes/kubernetes/issues/77310

Reply:
Since this is a general comment about `kubectl`, this is unrelated to `kubectl cp` the topic of this KEP, so this line should be removed.

Reply:
I've removed that.